src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/DiagnosticUtilities.cs

```diff
@@ -44,40 +50,110 @@ internal static string GetGenericParameterDeclaringMemberDisplayName(GenericPara
+            return true;
+        }
 
-            var decoded = ecmaMethod.GetDecodedCustomAttribute("System.Diagnostics.CodeAnalysis", requiresAttributeName);
-            if (decoded == null)
-                return null;
+        internal static string GetRequiresAttributeMessage(CustomAttributeValue<TypeDesc> attribute)
+        {
+            if (attribute.FixedArguments.Length != 0)
+                return (string)attribute.FixedArguments[0].Value;
 
-            var decodedValue = decoded.Value;
+            return null;
+        }
 
-            if (decodedValue.FixedArguments.Length != 0)
-                return (string)decodedValue.FixedArguments[0].Value;
+        internal static string GetRequiresAttributeUrl(CustomAttributeValue<TypeDesc> attribute)
+        {
+            if (attribute.NamedArguments.Length != 0 && attribute.NamedArguments[0].Name == "Url")
+                return (string)attribute.NamedArguments[0].Value;
 
             return null;
         }
 
-        internal static string GetRequiresAttributeUrl(MethodDesc method, string requiresAttributeName)
+        /// <summary>
+        /// Determines if method is within a declared Requires scope - this typically means that trim analysis
+        /// warnings should be suppressed in such a method.
+        /// </summary>
+        /// <remarks>Unlike <see cref="DoesMethodRequires(MethodDesc, string, out CustomAttributeValue?)"/>
+        /// if a declaring type has Requires, all methods in that type are considered "in scope" of that Requires. So this includes also
+        /// instance methods (not just statics and .ctors).</remarks>
+        internal static bool IsMethodInRequiresScope(this MethodDesc method, string requiresAttribute)
         {
-            var ecmaMethod = method.GetTypicalMethodDefinition() as EcmaMethod;
-            if (ecmaMethod == null)
-                return null;
+            if (method.HasCustomAttribute("System.Diagnostics.CodeAnalysis", requiresAttribute) && !method.IsStaticConstructor)
+                return true;
+
+            if (method.OwningType is TypeDesc type && TryGetRequiresAttribute(type, requiresAttribute, out _))
+                return true;
 
-            var decoded = ecmaMethod.GetDecodedCustomAttribute("System.Diagnostics.CodeAnalysis", requiresAttributeName);
-            if (decoded == null)
-                return null;
+            return false;
+        }
 
-            var decodedValue = decoded.Value;
+        /// <summary>
+		/// Determines if method requires (and thus any usage of such method should be warned about).
+		/// </summary>
+		/// <remarks>Unlike <see cref="IsMethodInRequiresScope(MethodDesc, string)"/> only static methods 
+		/// and .ctors are reported as requires when the declaring type has Requires on it.</remarks>
+		internal static bool DoesMethodRequires(this MethodDesc method, string requiresAttribute, [NotNullWhen(returnValue: true)] out CustomAttributeValue<TypeDesc>? attribute)
```

Comment:
Same comment as above linker uses this nameing scheme, analyzer doesn't. We should unify it.